x-pack/plugins/ml/server/models/data_recognizer/modules/apm_transaction/ml/datafeed_high_mean_transaction_duration.json

```diff
@@ -7,7 +7,7 @@
     "bool": {
       "filter": [
         { "term": { "processor.event": "transaction" } },
-        { "term": { "transaction.type": "request" } }
+        { "exists": { "field": "transaction.duration.us" } }
```

Comment:
Given the following request to create a new ML job:

```
POST /api/ml/modules/setup/apm_transaction
{
  // tagging the job with service.environment
  "custom_settings": {
    "service.environment": "production"
  },
  // tagging the job with app name
  "groups": ["apm"],
  // specifying the indicies to query
  "indexPatternName": "apm-*-transaction-*",
  // create job and start immediately
  "startDatafeed": true,
  // limit job to specific environment
  "query": {
    "bool": {
      "filter": [{ "term": { "service.environment": "production" } }]
    }
  }
}
```

Will the `query` specified by APM over the API and the `query` in the config file be merged? Or should we change the query to:

```
  "query": {
    "bool": {
      "filter": [
        { "term": { "processor.event": "transaction" } },
        { "exists": { "field": "transaction.duration.us" } }
        { "term": { "service.environment": "production" } }]
    }
  }
```
?

Reply:
Btw. overall does the request look correct?

Reply:
the `query` will not get merged, it will overwrite the query in each datafeed in the module.

here’s a suggestion for the setup request.
i’ve moved the custom settings into the `jobOverrides` and added a `job_tags` property. l’m not sold on that name, so better suggestions are welcome.
i’ve also removed the `groups` override because it’s not needed, all APM modules already have this group set.

```
POST /api/ml/modules/setup/apm_transaction
{
  // tagging the job with service.environment
  "jobOverrides": {
    "custom_settings": {
      "job_tags": {
        "service.environment": "production"
       }
     }
  },
  // specifying the indicies to query
  "indexPatternName": "apm-*-transaction-*",
  // create job and start immediately
  "startDatafeed": true,
  // limit job to specific environment
  "query": {
    "bool": {
      "filter": [
        { "term": { "processor.event": "transaction" } },
        { "exists": { "field": "transaction.duration.us" } }
        { "term": { "service.environment": "production" } }]
    }
  }
}
```

Reply:
Thanks @jgowdyelastic !

Is `jobOverrides` requried? I thought this would do:
```jsonc
POST /api/ml/modules/setup/apm_transaction
{
  "custom_settings": {
    "job_tags": {
      "service.environment": "production"
     }
   }
  //...
}
```

Reply:
yes, with it you can override any part of the job.

Reply:
Okay, so do we also need to do it for the other attributes like `query`?
```json
{
  "jobOverrides": {
    "query": {
      "bool": {
        "filter": [
          { "term": { "processor.event": "transaction" } },
          { "exists": { "field": "transaction.duration.us" } },
          { "term": { "service.environment": "production" } }
        ]
      }
    }
  }
}

```

Reply:
no, only for `custom_settings`.
`jobOverrides` and `datafeedOverrides` can be thought of as additional or advanced overrides for the job.
https://github.com/elastic/kibana/pull/42946

Reply:
Okay - that sounds a bit inconsistent to me. Is the intention to align this? Is it documented somewhere which fields need to be wrapped in `jobOverrides` and which don't?

Reply:
`query` and `groups` are the only overrides which could potentially be moved into `jobOverrides' and 'datafeedOverrides`. For the sake of backwards compatibility I'm happy for them to stay where there are, unless others agree that they are confusing and should be moved.